### DISCUSSION TITLE
Svc_GroundInterface unit tests

### DISCUSSION THREAD

**Discussion by JackNWhite** [nasa/fprime#1577] [Q&A]
Hello,

For reasons beyond my ken, our unit test scripts run the F' unit tests along with our own. I am getting wild running times for the Svc_GroundInterface unit test **Nominal.RandomizedGroundIf** - like 25 minutes, or 50 minutes.

Does anyone know why this might be the case? We have not changed the Svc_GroundInterface source.

**Comment by LeStarch** [comment]
https://github.com/nasa/fprime/blob/7cb9feed01d7be216ecbca6be0b1083a4a996d83/Svc/GroundInterface/test/ut/TestMain.cpp#L12

This is too much on slower systems.  I don't know why it would suddenly appear.  We should probably think about removing this component as it has been replaced with the broken-out Deframer and Framer components

**Reply by SterlingPeet** [reply]
@LeStarch : removing this component is tough on resource constrained systems.  Just an FYI, Arduino may need to move a version of it into the library repository to remain viable at that size.

**Reply by LeStarch** [reply]
@SterlingPeet, first, good to know. Is there something specific this component does differently?  Or is it just minimizing component count?

I am wondering if there is something that might be fixed with the replacement components, or if it is just the fact that there would be more components in that model.

**Comment by Anirban166** [comment]
As Michael hinted above, reducing the [step count](https://github.com/nasa/fprime/blob/7cb9feed01d7be216ecbca6be0b1083a4a996d83/Svc/GroundInterface/test/ut/TestMain.cpp#L12) (to something like `10` or even `1`) should be a solution. 

That would be the tl;dr of my post. If you're interested to know about what's going on underneath or why it takes longer than the other tests, read on as I try to explain from what I understand!

> For reasons beyond my ken, our unit test scripts run the F' unit tests along with our own. I am getting wild running times for the Svc_GroundInterface unit test Nominal.RandomizedGroundIf - like 25 minutes, or 50 minutes.
>
> Does anyone know why this might be the case? We have not changed the Svc_GroundInterface source.

In a nutshell, [RandomizedGroundIf](https://github.com/nasa/fprime/blob/7cb9feed01d7be216ecbca6be0b1083a4a996d83/Svc/GroundInterface/test/ut/TestMain.cpp#L17)<sup>1</sup> picks rules in a random sequence (from the four of them that have been declared and defined in the GroundInterface-specific [hpp](https://github.com/nasa/fprime/blob/master/Svc/GroundInterface/test/ut/GroundInterfaceRules.hpp) and [cpp](https://github.com/nasa/fprime/blob/master/Svc/GroundInterface/test/ut/GroundInterfaceRules.cpp) files respectively), runs them in correspondingly generated order, and repeats this process for `STEP_COUNT` times.

*What are 'rules'?* you might ask: 
Based on the [header](https://github.com/nasa/fprime/blob/master/STest/STest/Rule/Rule.hpp) which acts as an interface, I'd say rules are nothing but classes or structs that apart from having the presence of a constructor<sup>2</sup> call for initializing its name, must implement<sup>3</sup> two core functions that are declared in the header - the `precondition` (returns a boolean) and the `action` (the meat of the rule, containing logic that defines what it does). Both of them take the state<sup>4</sup> of the system by reference as input, and the action is executed (or the rule is applied, as given by the `apply` method from that header) if the precondition is met<sup>5</sup>. State modifications are made through this interface (given that the state of components shouldn't be altered directly), and the behaviours that these rules exhibit are being tested under this system.

<details>
<summary>Implementation-specific notes, as can be gathered from looking at the code</summary>
<sup>2</sup>Sets the passed value for the only member variable which is a string denoting the rule name, given by a char* with both the data and the pointer itself being made to be const-qualified immutable <br>
<sup>3</sup>They are pure virtual functions in the header file (Rules.hpp), thus requiring the subclasses (such as the deriving rule class) to override them <br>
<sup>4</sup>State is a user-defined type implemented as a class representing properties of the component, objects of which are being passed around here <br>
<sup>5</sup>This part is always valid here since the precondition implementations for all the rules return true
</details>

These rules are taken as input within certain '[Scenarios](https://github.com/nasa/fprime/tree/master/STest/STest/Scenario)' which at a higher level, plays with the ordering and repetitive execution of them for what's being used here (i.e. objects of `RandomScenario` and `BoundedScenario` that are being used in [TestMain.cpp](https://github.com/nasa/fprime/blob/master/Svc/GroundInterface/test/ut/TestMain.cpp)<sup>6</sup>).

That<sup>1</sup> particular test case establishes rule-based random testing wherein first a certain sequence of rules will be generated for the array of rules passed onto `RandomScenario`. Internally, it iterates over the rules array in a regular fashion but the rules get random indices allotted (limited by the total number of rules of course), and thus they are applied in that subsequently determined order. This class uses an `InterleavedScenario` in the process (which in turn takes an instance of `ScenarioArray` as an [argument](https://github.com/nasa/fprime/blob/7cb9feed01d7be216ecbca6be0b1083a4a996d83/STest/STest/Scenario/InterleavedScenario.hpp#L41) for its setter within the member initializer list of `InterleavedScenario`). This scenario makes a [call](https://github.com/nasa/fprime/blob/7cb9feed01d7be216ecbca6be0b1083a4a996d83/STest/STest/Scenario/InterleavedScenario.hpp#L81) to the method `getRandomIndex` which comes from `ScenarioArray`. That method makes a call to a function called `startLength`, which in turn invokes `inUnitInterval` (both of these being defined in [`Random.cpp`](https://github.com/nasa/fprime/blob/master/STest/STest/Random/Random.cpp)) where the random number generating function `bsd_random` is called (there is a `bsd_srandom` function too, which is the equivalent of seeded `bsd_random`, or what `srand` is to `rand`) at the end - that's quite the chain of functions we got here!

It is based on this rng that a random index gets generated for each rule in the scenario array, which leads to a randomized sequence of rules being applied, executing their action-based behaviour in that specific arrangement. In order to iteratively repeat this process for a desired number of times, a `BoundedScenario` that takes this `RandomScenario` is being used. 

At this point, I think the differences between this and the remaining three cases in the test runner code<sup>6</sup> for the GroundInterface component are clearly visible - the other tests just have one distinct fixed rule each, only test that particular behaviour once, and in general, don't involve going through much test code (like scenarios) or no randomization code at all if I'm not wrong (there is a rule called `RandomizeRule` that I see calling the functions from `Random/` within the action implementation for it, but that is not among the three cases that only test for the base rules separately). 
Compare this with `Nominal.RandomizedGroundIf`, which has to go through rule-scenario code with an rng phase for arbitrarily ordering four rules at once, has to test their individual behaviours in the determined sequence, and then has to repeat that `STEP_COUNT` times, which would definitely account for a large testing time for a large number of iterations. (this test should correctly consume 99.99% of the test runtime for a big `STEP_COUNT`)

That being said, the default value for the macro seems to be appropriate for present-day commercial computing devices as it takes less than a second for me to run those 10<sup>4</sup> rounds with all the setups that I'm using. For example, this is one benchmark for the default `STEP_COUNT` value from a container that feeds on a part of the resources from a base version of a 2019 (x86) 16" MBP: (still quite a bit of power compared to most embedded systems where fprime is tested ig, so might not be a good candidate to compare against)
 ```sh
~$ cd <fprime-checkout>/Svc/GroundInterface 
~/fprime/Svc/GroundInterface$ time fprime-util check >/dev/null

real	0m0.762s
user	0m0.573s
sys	0m0.193s
```
Meanwhile, running 10<sup>8</sup> rounds took me near about eleven minutes, which is reasonable given the number of iterations per second, but I'd definitely not wait that long either: (just tried this to experience a longer wait)
```sh
~/fprime/Svc/GroundInterface$ sed -i 's/#define STEP_COUNT 10000/#define STEP_COUNT 100000000/' test/ut/TestMain.cpp && time fprime-util check >/dev/null
 
real	10m56.805s
user	10m56.923s
sys	0m0.330s
```
Now while 10<sup>4</sup> seems rationable as a default value for such setups, I haven't been able to test anywhere else apart from my laptops, so I cannot comment if this would be accommodating for all use cases of fprime, but it does make sense for embedded systems with minimal computing power and resources to run these tests slower, in which case the most apparent solution would be to just reduce the step count.

**Comment by LeStarch** [comment]
Can we consider this question answered? Or is there more to explore here?
